### DISCUSSION TITLE
Grafana dashboard JSON => Jsonnet?

### DISCUSSION THREAD

**Discussion by jacksongoode** [grafana/grafonnet#32] [General]
I remember there was some discussion of a utility that would enable one to take the JSON model and translate into Jsonnet? @Duologic Is this still being developed or perhaps I could test a WIP? We are attempting to convert these JSON blobs into Jsonnet where there is a clearer structure and the ability to make use of Jsonnet's functionality.

**Comment by Duologic** [comment]
I've build an extremely naive first version for the internal hackathon, but the result is something that will be hard to maintain.

The main problem is that the JSON model that Grafana exposes today is very verbose, it includes a lot of default values. So rendering a Jsonnet version of this model will also be too verbose to be useful at this point as it doesn't know the difference between default values and values set by the user.

**Reply by jacksongoode** [reply]
Ah okay I see, yeah that was my thought that the benefit of a tool like this would be able to discern, through the schema, which are the non-default values for the dashboard.

Still, if this tool were hosted somewhere I would be interested in testing it out. It at least would do a lot of the initial heavy lifting 😃

**Reply by malcolmholmes** [reply]
And the issue with defaults is that, when we find a value in a json that matches the default, we don't know whether that means that it is a default, or that the user explicitly wants it set to that value. Were the default to change, we would not be able to deduce the intended value.

I'm not saying this cannot be overcome or worked around, but it certainly makes default removal harder to reason about.

**Reply by jacksongoode** [reply]
That makes sense. I think the strategy would be one where just plugging in a Grafana generated JSON model would provide a base template in jsonnet to start to cull away non-explicitly set parameters. This would of course have to be manual and with the understanding from the author as to what elements were actually modified (or which defaults ought to remain set) during the creation of the dashboard. I'm still hopeful that such a migration tool might exist with that context in mind.

**Reply by ekexium** [reply]
I'd love to try this out if possible. Modifying and reviewing .json files just kills me.

**Reply by jacksongoode** [reply]
Totally, the tedium is the hardest part of jumping on a new syntax.

**Comment by Duologic** [comment]
Secondary answer: Simply converting from JSON to Jsonnet has little benefit, the power is in the ability to make abstractions and reusable components that eventually compose one or more dashboards. For example, you'd define a Timeseries panel with a sensible configuration  (but without queries) and use it in multiple dashboards.

**Reply by jacksongoode** [reply]
I agree with the benefit being later. I think the conversion would simply serve the tedious task of translating JSON into the new Grafonnet syntax.

**Comment by jacksongoode** [comment]
Just want to follow up to see if this idea is possible now or manually constructing the dashboards from scratch is the recommendation? @Duologic

**Comment by DieBauer** [comment]
Late to the party. But I've found that Claude Opus 4.5 can do this work pretty OK. Your mileage may vary though ;)

**Comment by DanickBG** [comment]
A little bit late but for everyone else reading check here: [https://github.com/grafana/grafonnet/discussions/263](https://github.com/grafana/grafonnet/discussions/263)
